media/whip_server.go

```diff
@@ -0,0 +1,650 @@
+	// https://github.com/bluenviron/mediamtx/blob/4dfe274239a5a37198ce108250ae8db04f34cc3e/internal/protocols/whip/link_header.go#L24-L37
+	ret := make([]string, len(iceServers))
+
+	for i, server := range iceServers {
+		link := "<" + server.URLs[0] + ">; rel=\"ice-server\""
+		if server.Username != "" {
+			link += "; username=\"" + quoteCredential(server.Username) + "\"" +
+				"; credential=\"" + quoteCredential(server.Credential.(string)) + "\"; credential-type=\"password\""
+		}
+		ret[i] = link
+	}
+
+	return ret
+}
+
+func quoteCredential(v string) string {
+	b, _ := json.Marshal(v)
+	s := string(b)
+	return s[1 : len(s)-1]
+}
+
+func getUDPListenerAddr() (*net.UDPAddr, error) {
+	addrStr := os.Getenv("LIVE_AI_WHIP_ADDR") // TODO cli args?
+	if addrStr == "" {
+		// Default to all interfaces, port 7290
+		return &net.UDPAddr{
+			IP:   net.IP{0, 0, 0, 0},
+			Port: 7290,
+		}, nil
+	}
+
+	// Handle the ":PORT" shorthand notation
+	if strings.HasPrefix(addrStr, ":") {
+		port := 0
+		_, err := fmt.Sscanf(addrStr, ":%d", &port)
+		if err != nil {
+			return nil, fmt.Errorf("invalid UDP binding port: %v", err)
+		}
+		return &net.UDPAddr{
+			IP:   net.IP{0, 0, 0, 0},
+			Port: port,
+		}, nil
+	}
+
+	// Parse as a full IP:PORT address
+	udpAddr, err := net.ResolveUDPAddr("udp", addrStr)
+	if err != nil {
+		return nil, fmt.Errorf("invalid UDP address : %v", err)
+	}
+	return udpAddr, nil
+}
+
+func genParams() (func(*webrtc.API), func(*webrtc.API), func(*webrtc.API)) {
+	// Taken from Pion default codecs, but limited to to Opus and H.264
+
+	m := &webrtc.MediaEngine{}
+
+	// audio codecs
+	for _, codec := range []webrtc.RTPCodecParameters{
+		{
+			RTPCodecCapability: webrtc.RTPCodecCapability{webrtc.MimeTypeOpus, 48000, 2, "minptime=10;useinbandfec=1", nil},
+			PayloadType:        111,
+		},
+	} {
+		if err := m.RegisterCodec(codec, webrtc.RTPCodecTypeAudio); err != nil {
+			// this should really never happen
+			log.Fatal("could not register default codecs", err)
+		}
+	}
+
+	// video codecs
+	videoRTCPFeedback := []webrtc.RTCPFeedback{{"goog-remb", ""}, {"ccm", "fir"}, {"nack", ""}, {"nack", "pli"}}
+	for _, codec := range []webrtc.RTPCodecParameters{
+		{
+			RTPCodecCapability: webrtc.RTPCodecCapability{
+				webrtc.MimeTypeH264, 90000, 0,
+				"level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42001f",
+				videoRTCPFeedback,
+			},
+			PayloadType: 102,
+		},
+		{
+			RTPCodecCapability: webrtc.RTPCodecCapability{webrtc.MimeTypeRTX, 90000, 0, "apt=102", nil},
+			PayloadType:        103,
+		},
+
+		{
+			RTPCodecCapability: webrtc.RTPCodecCapability{
+				webrtc.MimeTypeH264, 90000, 0,
+				"level-asymmetry-allowed=1;packetization-mode=0;profile-level-id=42001f",
+				videoRTCPFeedback,
+			},
+			PayloadType: 104,
+		},
+		{
+			RTPCodecCapability: webrtc.RTPCodecCapability{webrtc.MimeTypeRTX, 90000, 0, "apt=104", nil},
+			PayloadType:        105,
+		},
+
+		// TODO verify 42e01f won't produle bframes
```

Comment:
Let's remove this before merging

Reply:
Sure, this was part of the default set of Pion codecs but I don't think this part even does what we need it to right now - the remote SDPs the browser is setting don't look as constrained as I was expecting. Needs more investigation